### PR TITLE
Refactor js

### PR DESCRIPTION
**Pull request by aomran**
This is some stuff I worked on during the flight back home. Again this was in two branches that I've combined together. With lots of JavaScript, the `domready.js` file became too big. It served us well for the 2-3 months that we used it but now it's time to move on.
#### New organization:
- As much of the logic and DOM manipulation should be  encapsulated into JavaScript objects. Each object should be in its own file inside of the `javascripts/objects` folder. By encapsulate I mean writing methods that clearly explain the intention of the code, making the code much more readable. 
- JavaScript that uses these objects is organized based on the controller name. For example: JavaScript associated with the `requesters` controller is in `requesters.js.coffee`  just like Rails default. The only exception is where the controller does too much. For example, the tracks controller JavaScript is split into a file for the student view and one for the teacher view.
- JavaScript that applies to the whole app is put inside the `application.js` file.
- This sort of follows the default Rails conventions, which now makes sense since we have sizeable JavaScript that we're working with.

The PR #109 introduced a lot of JavaScript and we have more to come. The app has been "progressively enhanced" so it still works without JavaScript. But the question needs to be asked, what baseline do we want to support?
- Support disabled JavaScript? This makes us keep a bunch of code that would be removed if we didn't need to support non-JS.
- IE 9+? I think we decided on this but not 100% sure.
- Mobile? I think we've done a good job so far of making this app mobile friendly.

### CODE REVIEW

**Comment by aomran** [comment]
This will require a visual check to make sure that all of the JavaScript works as intended. At some point I should add automated tests for the JavaScript.

**Comment by nachiketkumar** [comment]
Just started visually checking:
- drag drop not working
- show students who haven't voted toggle (eye icon) just refreshes page, doesnt show the hidden div.
- phase change select not working

**Comment by aomran** [comment]
The two commits above should fix the 3 issues that you listed. The analytics page js was just not merged properly because I got rid of `domread.js.coffee` without checking that everything was moved properly. The drag/drop issue is because of JavaScript's `this` bull which is confusing.

To explain the `this` confusion...

In coffeescript if you use `@` then that's a shortcut for `this.`. Inside of a `$().ready` or coffeescript's equivalent `$ ->`, `this` is the jQuery object. The `Curri` object is attached to window not jQuery. So the safest way to use the object is not to use `@` or `this.` since that would tell JavaScript to call the object called `Curri` which is directly inside jQuery. 

tl;dr ... JavaScript can be very confusing.

**Comment by nachiketkumar** [comment]
To respond to the questions in the original PR comment:
- I tend to think we don't support users who have disabled JS, but I'd like to know more about who these people might be. It has to be a tiny proportion of users, and with JS being so widespread, I don't feel too bad about this.
- IE9+ should be fine. MS is discontinuing support for IE8 in a [couple of months](http://zurb.com/article/1265/ie8-is-going-the-way-of-the-dodo-so-why-s), and pushing for auto updates as well. Amazingly, we have 1 IE student user (and 1 Safari, the rest being Chrome- no FF, I'm surprised).
- Agree about mobile, there are some small things that can be improved but overall I think it's a pretty decent experience.

**Comment by aomran** [comment]
Basecamp is an app that is not too JavaScript heavy but it just doesn't work at all without JavaScript. They don't even bother supporting non-js users. As for browsers, they have a page with information: https://basecamp.com/browsers

**Comment by aomran** [comment]
One final visual check and then this branch is ready to be merged. Things to check:
- Analytics: phase selector, "student hasn't voted box" show **(works)**
- Navigation: main-nav open/close + saves choice, sub-nav open/close, mobile-sidebar for tracks **(works)**
- Headers: student-help-status, teacher-request-number **(works)**\- _but position of the numbered label has moved up. I can figure out why this happened and fix._
- Invitations: Remove student from class **(works)**\- _student gets no notification of being removed though, and gets a rails error view if they are still logged in. On manually entering root_path, they get taken to classrooms view, in my case it was empty._
- Requesters: Remove student from help queue **(works)**\- *polling seems a little slow on the header- `.nav-icon.nav-help` not sure if it matters or not"
- Student Tracks: show success criteria, rate checkpoint **(works)**
- Teacher Tracks: date/time picker, delete checkpoint, sort checkpoints **(works)** - _we should add a message encouraging teachers to create checkpoints if a track has none in it, which will show only if there are no checkpoints. This can go into the minor issues branch._

Mixpanel events tracked directly in JavaScript:
- identify user (i.e. user's email, name, etc. are stored)
- Sign out
- Rate checkpoint
- Student Asked for Help
- Teacher Answered Student

_Look in the Mixpanel "Curri Development" app to see if these things are being tracked._

One thing I need to do more research about and work on is the best way of doing the above tests automatically to avoid the tedious manual tests.

**Comment by nachiketkumar** [comment]
I'll check the first lot, and let you guys check the mixpanel stuff.

**Comment by nachiketkumar** [comment]
ummm I don't know what happened but I seem to have pushed this branch again as a new pull request. I just made a small commit and wanted to push to this branch.

**Comment by aomran** [comment]
It's probably just a naming issue. If you did `$ git push origin refactorjs` instead of refactorJS. Just delete the remote branch and push again with the right name.

**Comment by nachiketkumar** [comment]
deleting `refactorjs` closed this pull request!

**Comment by aomran** [comment]
So this means we need to agree on naming conventions for the branches lol.

**Comment by nachiketkumar** [comment]
How we have it now is fine, I need to be more careful when setting up new branches to track..!

**Comment by pjfranzini** [comment]
I tested the mixpanel events listed. Everything is working great except sign-outs are still flaky (about half show up). Probably the least interesting of our events anyhow. Also, not on your list, but a bit of weird behaviour for sign-ins -- I am not having them show up till the "next person" signs in. Probably not an important issue either...

**Comment by aomran** [comment]
Ok this all looks good then. The following will be done/discussed in the minorchanges branch:
- Position of the numbered label for requesters badge is off.
- Students need to be notified of removal from class. _Note that in production no errors get showed, the user is directed to a page that says "We're sorry, but something went wrong (500)". This page can be found under the public folder and can be modified/styled to meet our needs. Btw, if the student was in only 1 class and they get removed, then the classrooms view will be empty (as it should)._
- Check polling speed
- Add a message encouraging teachers to create checkpoints if a track has none in it, which will show only if there are no checkpoints.
- Check sign-in and sign-out events again.

**Comment by aomran** [comment]
I've created a wiki entry for [CoffeeScript Organization](https://github.com/aomra015/curri/wiki/Coffee-script-Organization) and [Official Browser Support](https://github.com/aomra015/curri/wiki/Official-Browser-Support).

Merging this branch in a minute.
